src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,174 @@
+        {
+            Id = id;
+            Suggestions = suggestions;
+        }
+    }
+
+    /// <summary>
+    /// Provides a set of possible predictions for given input.
+    /// </summary>
+    public static class CommandPrediction
+    {
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the default timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens)
+        {
+            return PredictInput(ast, astTokens, millisecondsTimeout: 20);
+        }
+
+        /// <summary>
+        /// Collect the predictive suggestions from registered predictors using the specified timeout.
+        /// </summary>
+        /// <param name="ast">The <see cref="Ast"/> object from parsing the current command line input.</param>
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var cancellationSource = new CancellationTokenSource();
+            var cancellationToken = cancellationSource.Token;
+            var tasks = new Task<PredictionResult>[predictors.Count];
+
+            for (int i = 0; i < predictors.Count; i++)
+            {
+                var predictor = predictors[i];
+
+                tasks[i] = Task.Factory.StartNew(
+                    state => {
+                        var predictor = (IPredictor)state;
+                        List<string> texts = predictor.GetSuggestion(context, cancellationToken);
+                        return texts?.Count > 0 ? new PredictionResult(predictor.Id, texts) : null;
+                    }, predictor);
+            }
+
+            try
+            {
+                await Task.WhenAny(
+                    Task.WhenAll(tasks),
+                    Task.Delay(millisecondsTimeout, cancellationToken)).ConfigureAwait(false);
+                cancellationSource.Cancel();
+
+                var results = new List<PredictionResult>(predictors.Count);
+                foreach (Task<PredictionResult> task in tasks)
```

Comment:
Any task that isn't completed when the first task finished will be ignored, is that intended?

Reply:
I'm not sure I understand your question. Let me explain my intent here more clearly.
The code below is to make `PredictInput` stop waiting after timeout. `WhenAny` will return when either all `tasks` are completed or the time is up. 
```
                await Task.WhenAny(
                    Task.WhenAll(tasks),
                    Task.Delay(millisecondsTimeout, cancellationToken)).ConfigureAwait(false);
```

When code reaches `foreach (Task<PredictionResult> task in tasks)`, it's possible that
- some or even all of `tasks` are not yet completed;
- some or even all of `tasks` failed;
- some or even all of `tasks` succeeded;

For all of those possible cases, we only care about the `tasks` that `IsCompletedSuccessfully`. The rest are ignored intentionally.

Please let me know if that doesn't make sense, thanks!

Reply:
@daxian-dbw how are we intending to handle tasks that failed? I'm not seeing any handling for any failed tasks, nor anything that may be hanging indefinitely (should we have a timeout?).

That sounds kind of problematic for any kind of debugging or just general error handling that tasks that don't complete or error out never see the light of day again.

Reply:
> how are we intending to handle tasks that failed? I'm not seeing any handling for any failed tasks, 

From the perspective of `PredictInput`, a failed task is equivalent to "there is no prediction result from that predictor". It doesn't need to care why it fails and it shouldn't propagate any failures to the caller (_readline host_).

> nor anything that may be hanging indefinitely (should we have a timeout?).

A timeout is built in to make sure `Task.WhenAny` returns when the time is up.

As for an individual task, it may hang depending on what `predictor.GetSuggestion` does. `cancellationSource.Token` is passed to `predictor.GetSuggestion(context, cancellationSource.Token)` in the task, and when `Task.WhenAny` returns, `cancellationSource.Cancel()` is called to signal the cancellation to `predictor.GetSuggestion`, so if the predictor chooses to respect the cancellation token, it should be able to finish running.

> That sounds kind of problematic for any kind of debugging or just general error handling that tasks that don't complete or error out never see the light of day again.

Hope my explanation helps. It doesn't affect debugging as a task cannot be cancelled from outside, so when attaching to debugger, you can always see what `predicotr.GetSuggestion` is doing in case a hang or error happens.